Library/OcAppleKernelLib/MkextContext.c

```diff
@@ -0,0 +1,1040 @@
+  MkextVersion  = SwapBytes32 (MkextHeader->Common.Version);
+  NumKexts      = SwapBytes32 (MkextHeader->Common.NumKexts);
+
+  if (MkextHeader->Common.Magic != MKEXT_INVERT_MAGIC
+    || MkextHeader->Common.Signature != MKEXT_INVERT_SIGNATURE
+    || BufferSize != MkextSize
+    || OcOverflowAddU32 (NumKexts, NumReservedKexts, &NumMaxKexts)) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  //
+  // Mkext v1.
+  //
+  if (MkextVersion == MKEXT_VERSION_V1) {
+    if (OcOverflowMulAddU32 (sizeof (MKEXT_V1_KEXT), NumMaxKexts, sizeof (MKEXT_V1_HEADER), &CurrentOffset)) {
+      return EFI_INVALID_PARAMETER;
+    }
+    CurrentOffset = MKEXT_ALIGN (CurrentOffset);
+
+    if (MkextDecompress) {
+      //
+      // Copy header.
+      //
+      CopyMem (OutBuffer, Buffer, sizeof (MKEXT_V1_HEADER));
+      MkextHeaderOut = (MKEXT_HEADER_ANY *) OutBuffer;
+    }
+
+    //
+    // Process plists and binaries if present, decompressing as needed.
+    //
+    for (Index = 0; Index < NumKexts; Index++) {
+      PlistFullSize   = SwapBytes32 (MkextHeader->V1.Kexts[Index].Plist.FullSize);
+      BinFullSize     = SwapBytes32 (MkextHeader->V1.Kexts[Index].Binary.FullSize);
+
+      if (MkextDecompress) {
+        //
+        // Decompress entry.
+        //
+        PlistOffset     = SwapBytes32 (MkextHeader->V1.Kexts[Index].Plist.Offset);
+        PlistCompSize   = SwapBytes32 (MkextHeader->V1.Kexts[Index].Plist.CompressedSize);
+        BinOffset       = SwapBytes32 (MkextHeader->V1.Kexts[Index].Binary.Offset);
+        BinCompSize     = SwapBytes32 (MkextHeader->V1.Kexts[Index].Binary.CompressedSize);
+
+        if (CurrentOffset > OutBufferSize) {
+          return EFI_BUFFER_TOO_SMALL;
+        }
+
+        //
+        // Compressed size == 0 means no compression.
+        //
+        if (PlistCompSize > 0) {
+          if (OcOverflowAddU32 (PlistOffset, PlistCompSize, &Tmp)
+            || Tmp > MkextSize
+            || DecompressLZSS (
+              &OutBuffer[CurrentOffset],
+              PlistFullSize,
+              &((UINT8*)Buffer)[PlistOffset],
+              PlistCompSize
+            ) != PlistFullSize) {
+            return EFI_INVALID_PARAMETER;
+          }
+        } else {
+          if (OcOverflowAddU32 (PlistOffset, PlistFullSize, &Tmp)
+            || Tmp > MkextSize) {
+            return EFI_INVALID_PARAMETER;
+          }
+          CopyMem (&OutBuffer[CurrentOffset], &Buffer[PlistOffset], PlistFullSize);
+        }
+
+        MkextHeaderOut->V1.Kexts[Index].Plist.Offset          = SwapBytes32 (CurrentOffset);
+        MkextHeaderOut->V1.Kexts[Index].Plist.CompressedSize  = 0;
+        MkextHeaderOut->V1.Kexts[Index].Plist.FullSize        = SwapBytes32 (PlistFullSize);
+        MkextHeaderOut->V1.Kexts[Index].Plist.ModifiedSeconds = MkextHeader->V1.Kexts[Index].Plist.ModifiedSeconds;
+        if (OcOverflowAddU32 (CurrentOffset, MKEXT_ALIGN (PlistFullSize), &CurrentOffset)) {
```

Comment:
This looks wrong. The memory was already copied at `CurrentOffset`, checking for overflow here means that we have already corrupted the memory. This needs to become a simple addition without the overflow maths.